File: scripts/algolia-related-docs.js
// generate-algolia-related.js
// Enhanced with doc_type intelligence and improved semantic matching
// Updated for modern Algolia API and better error handling
// Modified to get all results and merge to page-level entries

const { algoliasearch } = require('algoliasearch');
const fs = require('fs').promises;
const path = require('path');
require('dotenv').config(); // Add this dependency: npm install dotenv

// Use environment variables for security
const APP_ID = process.env.ALGOLIA_APP_ID || '7AL1W7YVZK';
const API_KEY = process.env.ALGOLIA_API_KEY;
const INDEX_NAME = process.env.ALGOLIA_INDEX_NAME || 'clickhouse';

if (!API_KEY) {
  console.error('❌ ALGOLIA_API_KEY environment variable is required');
  process.exit(1);
}

// Updated for Algolia v5+ API
const client = algoliasearch(APP_ID, API_KEY);

// doc_type priority scores (matching your indexing script)
const DOC_TYPE_SCORES = {
  'tutorial': 1000,
  'how-to': 900,
  'explanation': 800,
  'overview': 700,
  'reference': 100
};

// Rate limiting helper
class RateLimiter {
  constructor(requestsPerSecond = 10) {
    this.requestsPerSecond = requestsPerSecond;
    this.interval = 1000 / requestsPerSecond;
    this.lastRequestTime = 0;
  }

  async waitIfNeeded() {
    const now = Date.now();
    const timeSinceLastRequest = now - this.lastRequestTime;
    
    if (timeSinceLastRequest < this.interval) {
      const waitTime = this.interval - timeSinceLastRequest;
      await new Promise(resolve => setTimeout(resolve, waitTime));
    }
    
    this.lastRequestTime = Date.now();
  }
}

const rateLimiter = new RateLimiter(8); // 8 requests per second to be safe

// Function to merge section-level docs into page-level docs
function mergeToPageLevel(allDocs) {
  console.log('🔄 Merging section-level entries into page-level documents...');
  
  const pageMap = new Map();
  
  allDocs.forEach(doc => {
    if (!doc.url) return;
    
    // Get the base URL without hash fragment
    const baseUrl = doc.url.split('#')[0];
    
    if (!pageMap.has(baseUrl)) {
      // Create a new page entry
      const pageDoc = {
        objectID: doc.objectID,
        title: doc.title || '',
        content: doc.content || '',
        url: baseUrl,
        type: 'page', // Mark as merged page-level doc
        lvl0: doc.lvl0 || '',
        lvl1: doc.lvl1 || '',
        doc_type: doc.doc_type || 'reference',
        doc_type_score: doc.doc_type_score || DOC_TYPE_SCORES[doc.doc_type] || 100,
        keywords: doc.keywords || '',
        sections: [] // Track what sections were merged
      };
      
      pageMap.set(baseUrl, pageDoc);
    }
    
    const pageDoc = pageMap.get(baseUrl);
    
    // Merge content from sections
    if (doc.content && !pageDoc.content.includes(doc.content)) {
      pageDoc.content += ' ' + doc.content;
    }
    
    // Merge keywords
    if (doc.keywords && !pageDoc.keywords.includes(doc.keywords)) {
      pageDoc.keywords += (pageDoc.keywords ? ', ' : '') + doc.keywords;
    }
    
    // Track sections that were merged
    if (doc.url.includes('#')) {
      pageDoc.sections.push({
        title: doc.title,
        type: doc.type,
        url: doc.url
      });
    }
    
    // Prefer page-level titles over section titles
    if (!doc.url.includes('#') && doc.title) {
      pageDoc.title = doc.title;
      pageDoc.objectID = doc.objectID; // Use page-level objectID
    }
    
    // Use the highest priority doc_type found
    const currentScore = DOC_TYPE_SCORES[pageDoc.doc_type] || 100;
    const newScore = DOC_TYPE_SCORES[doc.doc_type] || 100;
    if (newScore > currentScore) {
      pageDoc.doc_type = doc.doc_type;
      pageDoc.doc_type_score = newScore;
    }
  });
  
  const mergedDocs = Array.from(pageMap.values());
  
  console.log(`  📄 Merged ${allDocs.length} entries into ${mergedDocs.length} page-level documents`);
  
  // Show merge statistics
  const sectionStats = {};
  mergedDocs.forEach(doc => {
    const sectionCount = doc.sections.length;
    sectionStats[sectionCount] = (sectionStats[sectionCount] || 0) + 1;
  });
  
  console.log('  📊 Sections per page distribution:', sectionStats);
  
  return mergedDocs;
}

async function getAllDocsWithRetry(maxRetries = 3) {
  console.log('📚 Getting all docs using efficient multi-query approach...');
  console.log('💰 Estimating ~10 API calls for discovery');
  
  const allHits = [];
  const seenObjectIDs = new Set();
  
  const queries = [
    { query: '', hitsPerPage: 1000 }, // Get the main 1000
    { filters: 'url:*/sql-reference/*', hitsPerPage: 200 },
    { filters: 'url:*/cloud/*', hitsPerPage: 200 },
    { filters: 'url:*/integrations/*', hitsPerPage: 200 },
    { filters: 'url:*/guides/*', hitsPerPage: 200 },
    { filters: 'url:*/chdb/*', hitsPerPage: 200 },
    { filters: 'type:lvl1', hitsPerPage: 500 },
    { filters: 'type:lvl2', hitsPerPage: 500 },
    { query: 'reference', hitsPerPage: 300 },
    { query: 'tutorial guide', hitsPerPage: 300 }
  ];
  
  for (const [index, searchParams] of queries.entries()) {
    let retries = 0;
    let success = false;
    
    while (!success && retries < maxRetries) {
      try {
        await rateLimiter.waitIfNeeded();
        
        const response = await client.searchSingleIndex({
          indexName: INDEX_NAME,
          searchParams: {
            ...searchParams,
            attributesToRetrieve: [
              'objectID', 'title', 'content', 'url', 'type', 'lvl0', 'lvl1',
              'doc_type', 'doc_type_score', 'keywords', 'h1', 'h2', 'h3'
            ]
          }
        });
        
        let newHits = 0;
        for (const hit of response.hits) {
          if (!seenObjectIDs.has(hit.objectID)) {
            seenObjectIDs.add(hit.objectID);
            allHits.push(hit);
            newHits++;
          }
        }
        
        console.log(`  ✅ Query ${index + 1}: ${newHits} new docs (total: ${allHits.length})`);
        success = true;
        
      } catch (error) {
        retries++;
        console.warn(`  ⚠️ Query ${index + 1} failed (attempt ${retries}/${maxRetries}):`, error.message);
        
        if (retries < maxRetries) {
          const backoffTime = Math.pow(2, retries) * 1000; // Exponential backoff
          console.log(`    Retrying in ${backoffTime / 1000}s...`);
          await new Promise(resolve => setTimeout(resolve, backoffTime));
        }
      }
    }
    
    if (!success) {
      console.error(`  ❌ Query ${index + 1} failed after ${maxRetries} attempts`);
    }
  }
  
  console.log(`\n🎉 Raw result: ${allHits.length} unique documents`);
  
  // Now merge to page-level
  const mergedDocs = mergeToPageLevel(allHits);
  
  return mergedDocs;
}

function calculateDocTypeCompatibility(sourceDoc, targetDoc) {
  const sourceType = sourceDoc.doc_type || 'reference';
  const targetType = targetDoc.doc_type || 'reference';
  
  // Same doc_type gets highest bonus
  if (sourceType === targetType) {
    return 100;
  }
  
  // Compatible doc_type combinations
  const compatibilityMap = {
    'tutorial': { 'how-to': 80, 'explanation': 60, 'overview': 40, 'reference': 20 },
    'how-to': { 'tutorial': 80, 'explanation': 70, 'overview': 30, 'reference': 25 },
    'explanation': { 'tutorial': 60, 'how-to': 70, 'overview': 50, 'reference': 30 },
    'overview': { 'tutorial': 40, 'how-to': 30, 'explanation': 50, 'reference': 60 },
    'reference': { 'tutorial': 20, 'how-to': 25, 'explanation': 30, 'overview': 60 }
  };
  
  return compatibilityMap[sourceType]?.[targetType] || 10;
}

function calculateKeywordSimilarity(sourceDoc, targetDoc) {
  const sourceKeywords = parseKeywords(sourceDoc.keywords);
  const targetKeywords = parseKeywords(targetDoc.keywords);
  
  if (sourceKeywords.length === 0 || targetKeywords.length === 0) {
    return 0;
  }
  
  const intersection = sourceKeywords.filter(k => 
    targetKeywords.some(tk => tk.includes(k) || k.includes(tk))
  );
  
  return (intersection.length / Math.max(sourceKeywords.length, targetKeywords.length)) * 100;
}

function parseKeywords(keywordString) {
  if (!keywordString || typeof keywordString !== 'string') return [];
  
  return keywordString
    .split(',')
    .map(k => k.trim().toLowerCase())
    .filter(k => k.length > 2); // Filter out very short keywords
}

function isDuplicateContent(sourceDoc, targetDoc) {
  // Compare base URLs without fragments since we're dealing with page-level docs
  if (sourceDoc.url && targetDoc.url) {
    const sourceBase = sourceDoc.url.split('#')[0];
    const targetBase = targetDoc.url.split('#')[0];
    return sourceBase === targetBase;
  }
  return false;
}

function prepareSearchQuery(doc) {
  let searchText = doc.content || doc.title || '';
  
  if (!searchText.trim()) {
    return '';
  }
  
  // Clean up the content first
  searchText = searchText
    .replace(/[`*_#\[\](){}]/g, ' ') // Remove markdown formatting
    .replace(/https?:\/\/[^\s]+/g, ' ') // Remove URLs
    .replace(/\s+/g, ' ') // Normalize whitespace
    .trim();
  
  // Smart truncation - keep complete sentences
  if (searchText.length > 300) {
    const truncated = searchText.substring(0, 300);
    const lastSentence = truncated.lastIndexOf('.');
    const lastSpace = truncated.lastIndexOf(' ');
    
    if (lastSentence > 150) {
      searchText = truncated.substring(0, lastSentence + 1);
    } else if (lastSpace > 150) {
      searchText = truncated.substring(0, lastSpace);
    } else {
      searchText = truncated;
    }
  }
  
  // Add top keywords if there's space
  if (doc.keywords && searchText.length < 250) {
    const keywords = parseKeywords(doc.keywords).slice(0, 3);
    const keywordText = keywords.join(' ');
    searchText += ` ${keywordText}`;
  }
  
  // Final length check (Algolia query limit is around 512 bytes)
  if (searchText.length > 400) {
    searchText = searchText.substring(0, 400);
  }
  
  return searchText.trim();
}

async function findRelatedDocs(doc, allDocs, maxRetries = 3) {
  let retries = 0;
  
  while (retries < maxRetries) {
    try {
      await rateLimiter.waitIfNeeded();
      
      const searchText = prepareSearchQuery(doc);
      
      if (!searchText) {
        console.warn(`⚠️ Empty search text for ${doc.objectID}, using title only`);
        const fallbackQuery = doc.title?.substring(0, 100) || 'untitled';
        if (!fallbackQuery || fallbackQuery === 'untitled') {
          return [];
        }
      }
      
      // Search against the original Algolia index (not our merged docs)
      const response = await client.searchSingleIndex({
        indexName: INDEX_NAME,
        searchParams: {
          query: searchText || doc.title?.substring(0, 100),
          hitsPerPage: 20, // Get more results since we'll merge them
          filters: `NOT objectID:${doc.objectID}`,
          attributesToRetrieve: [
            'objectID', 'title', 'url', 'type', 'lvl0', 'lvl1', 
            'doc_type', 'doc_type_score', 'keywords', 'content'
          ],
          typoTolerance: true,
          ignorePlurals: true
        }
      });
      
      // Merge the search results to page-level before ranking
      const mergedResults = mergeToPageLevel(response.hits);
      
      // Enhanced filtering and ranking
      const candidates = mergedResults
        .filter(hit => {
          if (hit.objectID === doc.objectID || hit.title === doc.title) return false;
          if (isDuplicateContent(doc, hit)) return false;
          return true;
        })
        .map((hit, index) => {
          const docTypeBonus = calculateDocTypeCompatibility(doc, hit);
          const keywordSimilarity = calculateKeywordSimilarity(doc, hit);
          const positionScore = Math.max(0, 100 - (index * 5)); // Position in search results
          
          // Combined score with weights
          const totalScore = (
            positionScore * 0.4 +           // Algolia's ranking
            docTypeBonus * 0.35 +          // doc_type compatibility  
            keywordSimilarity * 0.25       // Keyword overlap
          );
          
          return {
            ...hit,
            enhanced_score: Math.round(totalScore),
            doc_type_bonus: docTypeBonus,
            keyword_similarity: Math.round(keywordSimilarity),
            position_score: positionScore
          };
        })
        .sort((a, b) => b.enhanced_score - a.enhanced_score)
        .slice(0, 5)
        .map((hit) => ({
          id: hit.url?.replace('https://clickhouse.com/docs/', '').replace(/\/$/, '') || hit.objectID,
          title: hit.title || hit.objectID || 'Untitled',
          url: hit.url,
          type: hit.type,
          doc_type: hit.doc_type || 'reference',
          reason: generateReason(hit, doc),
          similarity_score: hit.enhanced_score,
          sections_merged: hit.sections?.length || 0,
          debug: {
            doc_type_bonus: hit.doc_type_bonus,
            keyword_similarity: hit.keyword_similarity,
            position_score: hit.position_score
          }
        }));
        
      return candidates;
      
    } catch (error) {
      retries++;
      console.warn(`⚠️ Error finding related docs for ${doc.objectID} (attempt ${retries}/${maxRetries}):`, error.message);
      
      if (retries < maxRetries) {
        const backoffTime = Math.pow(2, retries) * 1000;
        await new Promise(resolve => setTimeout(resolve, backoffTime));
      }
    }
  }
  
  console.error(`❌ Failed to find related docs for ${doc.objectID} after ${maxRetries} attempts`);
  return [];
}

function generateReason(relatedDoc, sourceDoc) {
  const reasons = [];
  
  if (relatedDoc.doc_type === sourceDoc.doc_type) {
    const typeNames = {
      'tutorial': 'tutorial',
      'how-to': 'how-to guide', 
      'explanation': 'explanation',
      'overview': 'overview',
      'reference': 'reference'
    };
    reasons.push(`Same type (${typeNames[sourceDoc.doc_type] || sourceDoc.doc_type})`);
  }
  
  if (relatedDoc.keyword_similarity > 30) {
    reasons.push('Similar topics');
  }
  
  if (relatedDoc.doc_type_bonus > 60) {
    reasons.push('Compatible content');
  }
  
  if (relatedDoc.sections_merged > 3) {
    reasons.push('Comprehensive page');
  }
  
  return reasons.length > 0 ? reasons.join(', ') : `Related content (${relatedDoc.enhanced_score}% match)`;
}

function isNavigationPage(doc) {
  if (!doc.title || !doc.url) return true;
  
  const navigationPatterns = [
    /^index$/i,
    /overview$/i,
    /introduction$/i,
    /getting.started$/i,
    /^intro$/i,
    /concepts$/i
  ];
  
  const docId = doc.url.replace('/docs/', '').replace('/', '');
  return navigationPatterns.some(pattern => pattern.test(docId)) ||
         doc.title.toLowerCase().includes('overview') ||
         doc.title.toLowerCase().includes('index');
}

async function generateRelatedDocs() {
  try {
    console.log('🚀 Starting enhanced Algolia-powered related docs generation (PAGE-LEVEL MERGING)...\n');
    
    // Get all docs and merge to page-level
    const allDocs = await getAllDocsWithRetry();
    
    if (allDocs.length === 0) {
      throw new Error('No documents retrieved from Algolia');
    }
    
    // Filter for quality content
    const contentDocs = allDocs.filter(doc => 
      doc.objectID && 
      (doc.title || doc.url) &&
      !isNavigationPage(doc)
    );
    
    console.log(`📖 Processing ${contentDocs.length} page-level docs (filtered from ${allDocs.length} merged)`);
    
    // Show doc_type distribution
    const docTypeStats = {};
    contentDocs.forEach(doc => {
      const type = doc.doc_type || 'unknown';
      docTypeStats[type] = (docTypeStats[type] || 0) + 1;
    });
    
    console.log(`📊 Doc type distribution:`, docTypeStats);
    
    // Show sections merged distribution
    const sectionStats = {};
    contentDocs.forEach(doc => {
      const count = doc.sections?.length || 0;
      const bucket = count === 0 ? '0' : count <= 3 ? '1-3' : count <= 10 ? '4-10' : '10+';
      sectionStats[bucket] = (sectionStats[bucket] || 0) + 1;
    });
    
    console.log(`📄 Sections merged per page:`, sectionStats);
    
    const relatedDocsMap = {};
    let processed = 0;
    let errors = 0;
    
    // Process docs in smaller batches with better error handling
    const batchSize = 5; // Reduced batch size for better rate limiting
    for (let i = 0; i < contentDocs.length; i += batchSize) {
      const batch = contentDocs.slice(i, i + batchSize);
      
      const batchPromises = batch.map(async (doc) => {
        try {
          const docId = doc.url?.replace('https://clickhouse.com/docs/', '').replace(/\/$/, '') || doc.objectID;
          const relatedDocs = await findRelatedDocs(doc, contentDocs);
          relatedDocsMap[docId] = relatedDocs;
          
          processed++;
          if (processed % 25 === 0) {
            console.log(`⏳ Processed ${processed}/${contentDocs.length} page-level docs (${errors} errors)...`);
          }
        } catch (error) {
          errors++;
          console.error(`❌ Failed to process doc ${doc.objectID}:`, error.message);
        }
      });
      
      await Promise.all(batchPromises);
      
      // Longer delay between batches for rate limiting
      if (i + batchSize < contentDocs.length) {
        await new Promise(resolve => setTimeout(resolve, 500));
      }
    }
    
    // Save results with error handling
    const outputPath = 'src/data/related-docs.json';
    try {
      await fs.mkdir(path.dirname(outputPath), { recursive: true });
      await fs.writeFile(outputPath, JSON.stringify(relatedDocsMap, null, 2));
      console.log(`💾 Saved to: ${outputPath}`);
    } catch (error) {
      console.error('❌ Failed to save results:', error);
      throw error;
    }
    
    // Enhanced statistics
    const docsWithRelated = Object.values(relatedDocsMap).filter(related => related.length > 0);
    const avgRelated = docsWithRelated.length > 0 
      ? docsWithRelated.reduce((sum, related) => sum + related.length, 0) / docsWithRelated.length 
      : 0;
    
    console.log(`\n📊 Final Statistics (PAGE-LEVEL MERGED):`);
    console.log(`  - Total page-level docs processed: ${processed}`);
    console.log(`  - Pages with related content: ${docsWithRelated.length}`);
    console.log(`  - Average related docs per page: ${avgRelated.toFixed(1)}`);
    console.log(`  - Errors encountered: ${errors}`);
    console.log(`  - Success rate: ${((processed - errors) / processed * 100).toFixed(1)}%`);
    
    // Show doc_type matching stats
    const typeMatchStats = {};
    Object.values(relatedDocsMap).forEach(related => {
      related.forEach(r => {
        typeMatchStats[r.doc_type] = (typeMatchStats[r.doc_type] || 0) + 1;
      });
    });
    console.log(`  - Related docs by type:`, typeMatchStats);
    
    // Show examples with enhanced scoring
    const examples = Object.entries(relatedDocsMap)
      .filter(([_, related]) => related.length > 0)
      .slice(0, 3);
      
    if (examples.length > 0) {
      console.log(`\n🔍 Examples with enhanced scoring (PAGE-LEVEL MERGED):`);
      examples.forEach(([docId, related]) => {
        console.log(`\n  ${docId}:`);
        related.slice(0, 3).forEach(r => {
          console.log(`    → ${r.title} (${r.similarity_score}% match, ${r.doc_type}, ${r.sections_merged} sections)`);
          console.log(`      Reason: ${r.reason}`);
        });
      });
    }
    
    console.log(`\n🎉 Done! Your related docs now use merged page-level entries with intelligent doc_type matching!`);
    
  } catch (error) {
    console.error('❌ Fatal error generating related docs:', error);
    process.exit(1);
  }
}

// Run the generator
if (require.main === module) {
  generateRelatedDocs();
}

module.exports = { generateRelatedDocs };

File: scripts/algolia-related-docs.py
#!/usr/bin/env python3
"""
Related docs generator for Algolia.
Generates relationships between ClickHouse documentation pages using local similarity scoring.
"""

import argparse
import json
import os
import time
from algoliasearch.search.client import SearchClientSync


# Configuration
CONFIG = {
    'MIN_SIMILARITY_SCORE': 20,
    'MAX_RELATED_DOCS': 5,
    'RATE_LIMIT_DELAY': 0.1,
}

# Doc type compatibility matrix
DOC_TYPE_COMPATIBILITY = {
    'tutorial': {'how-to': 80, 'explanation': 60, 'overview': 40, 'reference': 20},
    'how-to': {'tutorial': 80, 'explanation': 70, 'overview': 30, 'reference': 25},
    'explanation': {'tutorial': 60, 'how-to': 70, 'overview': 50, 'reference': 30},
    'overview': {'tutorial': 40, 'how-to': 30, 'explanation': 50, 'reference': 60},
    'reference': {'tutorial': 20, 'how-to': 25, 'explanation': 30, 'overview': 60}
}


def rate_limit_sleep():
    """Simple rate limiting to avoid hitting Algolia limits"""
    time.sleep(CONFIG['RATE_LIMIT_DELAY'])


def safe_get_attr(hit, attr, default=''):
    """Safely get attribute from Algolia Hit object"""
    try:
        # Try standard attribute access
        if hasattr(hit, attr):
            val = getattr(hit, attr, default)
            if val is not None:
                return val
        
        # Try dictionary-style access
        try:
            val = hit[attr]
            if val is not None:
                return val
        except (KeyError, TypeError):
            pass
        
        # For objectID, try common alternatives
        if attr == 'objectID':
            for alt_name in ['object_id', 'id', '_id', 'objectId']:
                if hasattr(hit, alt_name):
                    val = getattr(hit, alt_name, default)
                    if val is not None:
                        return val
                try:
                    val = hit[alt_name]
                    if val is not None:
                        return val
                except (KeyError, TypeError):
                    continue
        
        return default
    except Exception:
        return default


def get_all_docs(client, index_name):
    """Get all documents using multiple targeted queries for comprehensive coverage"""
    print("📚 Fetching documents from Algolia...")
    
    all_docs = {}  # Use dict to deduplicate by objectID
    
    # Multiple queries to ensure comprehensive coverage
    queries = [
        {'query': '', 'hitsPerPage': 1000},  # Main empty query
        {'filters': 'url:*/sql-reference/*', 'hitsPerPage': 300},
        {'filters': 'url:*/integrations/*', 'hitsPerPage': 300}, 
        {'filters': 'url:*/cloud/*', 'hitsPerPage': 200},
        {'filters': 'url:*/guides/*', 'hitsPerPage': 200},
        {'query': 'tutorial', 'hitsPerPage': 200},
        {'query': 'reference', 'hitsPerPage': 200},
        {'query': 'how-to', 'hitsPerPage': 200},
        {'query': 'clickhouse', 'hitsPerPage': 300}
    ]
    
    for i, search_params in enumerate(queries, 1):
        rate_limit_sleep()
        try:
            response = client.search_single_index(
                index_name=index_name,
                search_params={
                    **search_params,
                    'attributesToRetrieve': [
                        'objectID', 'title', 'url', 'doc_type', 'keywords', 
                        'content', 'hierarchy', 'type'
                    ]
                }
            )
            
            new_count = 0
            for hit in response.hits:
                object_id = safe_get_attr(hit, 'objectID')
                
                if object_id and object_id not in all_docs:
                    all_docs[object_id] = hit
                    new_count += 1
            
            print(f"  Query {i}: {new_count} new docs, {len(all_docs)} total")
            
        except Exception as e:
            print(f"  Query {i} failed: {e}")
    
    docs_list = list(all_docs.values())
    print(f"✅ Got {len(docs_list)} unique documents")
    return docs_list


def calculate_similarity_score(source_hit, target_hit):
    """Calculate similarity score between two documents"""
    
    # Doc type compatibility (35% weight)
    source_type = safe_get_attr(source_hit, 'doc_type', 'reference')
    target_type = safe_get_attr(target_hit, 'doc_type', 'reference')
    
    if source_type == target_type:
        doc_type_score = 100
    else:
        doc_type_score = DOC_TYPE_COMPATIBILITY.get(source_type, {}).get(target_type, 10)
    
    # Keyword similarity (25% weight)
    keyword_score = calculate_keyword_similarity(source_hit, target_hit)
    
    # Title similarity (40% weight)
    title_score = calculate_title_similarity(source_hit, target_hit)
    
    # Combined weighted score
    total_score = (doc_type_score * 0.35) + (keyword_score * 0.25) + (title_score * 0.40)
    return round(total_score)


def calculate_keyword_similarity(source_hit, target_hit):
    """Calculate keyword overlap similarity"""
    source_keywords = safe_get_attr(source_hit, 'keywords', '')
    target_keywords = safe_get_attr(target_hit, 'keywords', '')
    
    source_kw_set = set()
    target_kw_set = set()
    
    if source_keywords:
        source_kw_set = {k.strip().lower() for k in source_keywords.split(',') 
                        if k.strip() and len(k.strip()) > 2}
    if target_keywords:
        target_kw_set = {k.strip().lower() for k in target_keywords.split(',') 
                        if k.strip() and len(k.strip()) > 2}
    
    if source_kw_set and target_kw_set:
        overlap = len(source_kw_set & target_kw_set)
        total = len(source_kw_set | target_kw_set)
        return (overlap / total) * 100 if total > 0 else 0
    else:
        return 0


def calculate_title_similarity(source_hit, target_hit):
    """Calculate title word overlap similarity"""
    source_title = safe_get_attr(source_hit, 'title', '')
    target_title = safe_get_attr(target_hit, 'title', '')
    
    source_words = {w.lower() for w in source_title.split() if len(w) > 3}
    target_words = {w.lower() for w in target_title.split() if len(w) > 3}
    
    if source_words and target_words:
        overlap = len(source_words & target_words)
        total = len(source_words | target_words)
        return (overlap / total) * 100 if total > 0 else 0
    else:
        return 0


def find_related_docs(source_hit, all_hits):
    """Find related documents for a source document"""
    candidates = []
    source_object_id = safe_get_attr(source_hit, 'objectID')
    
    for target_hit in all_hits:
        target_object_id = safe_get_attr(target_hit, 'objectID')
        target_title = safe_get_attr(target_hit, 'title')
        target_url = safe_get_attr(target_hit, 'url')
        
        # Skip self and documents without proper data
        if (target_object_id == source_object_id or
            not target_title or
            not target_url):
            continue
        
        # Calculate similarity
        score = calculate_similarity_score(source_hit, target_hit)
        
        # Only keep documents with reasonable similarity
        if score > CONFIG['MIN_SIMILARITY_SCORE']:
            candidates.append({
                'hit': target_hit,
                'score': score
            })
    
    # Sort by score and take top results
    candidates.sort(key=lambda x: x['score'], reverse=True)
    candidates = candidates[:CONFIG['MAX_RELATED_DOCS']]
    
    # Format results for JSON output
    return format_related_docs(source_hit, candidates)


def format_related_docs(source_hit, candidates):
    """Format candidate documents for JSON output"""
    related = []
    source_doc_type = safe_get_attr(source_hit, 'doc_type', 'reference')
    
    for candidate in candidates:
        hit = candidate['hit']
        
        hit_doc_type = safe_get_attr(hit, 'doc_type', 'reference')
        hit_title = safe_get_attr(hit, 'title')
        hit_url = safe_get_attr(hit, 'url')
        
        # Generate reason
        reason = generate_reason(source_doc_type, hit_doc_type, candidate['score'])
        
        # Clean URL for ID
        doc_id = hit_url.replace('https://clickhouse.com/docs/', '').strip('/')
        
        related.append({
            'id': doc_id,
            'title': hit_title,
            'url': hit_url,
            'doc_type': hit_doc_type,
            'reason': reason,
            'similarity_score': candidate['score']
        })
    
    return related


def generate_reason(source_doc_type, target_doc_type, score):
    """Generate a human-readable reason for the relationship"""
    if target_doc_type == source_doc_type:
        return f"Same type ({target_doc_type})"
    
    doc_bonus = DOC_TYPE_COMPATIBILITY.get(source_doc_type, {}).get(target_doc_type, 0)
    if doc_bonus > 60:
        return "Compatible content"
    
    return f"Related content ({score}% match)"


def filter_valid_hits(all_hits):
    """Filter to hits that have the required fields"""
    valid_hits = []
    for hit in all_hits:
        if (safe_get_attr(hit, 'objectID') and 
            safe_get_attr(hit, 'title') and 
            safe_get_attr(hit, 'url')):
            valid_hits.append(hit)
    
    return valid_hits


def show_doc_type_distribution(valid_hits):
    """Show distribution of document types"""
    doc_types = {}
    for hit in valid_hits:
        dt = safe_get_attr(hit, 'doc_type', 'unknown')
        doc_types[dt] = doc_types.get(dt, 0) + 1
    
    print(f"📊 Doc types: {doc_types}")


def generate_related_docs_map(valid_hits):
    """Generate the complete related docs mapping"""
    related_docs_map = {}
    
    for i, hit in enumerate(valid_hits):
        try:
            # Create doc ID from URL
            hit_url = safe_get_attr(hit, 'url')
            doc_id = hit_url.replace('https://clickhouse.com/docs/', '').strip('/')
            
            # Find related docs
            related = find_related_docs(hit, valid_hits)
            related_docs_map[doc_id] = related
            
            if (i + 1) % 50 == 0:
                print(f"⏳ Processed {i + 1}/{len(valid_hits)} documents...")
                
        except Exception as e:
            hit_title = safe_get_attr(hit, 'title', 'unknown')
            print(f"❌ Error processing {hit_title}: {e}")
    
    return related_docs_map


def save_results(related_docs_map, output_path):
    """Save results to JSON file"""
    print(f"💾 Saving to {output_path}...")
    os.makedirs(os.path.dirname(output_path), exist_ok=True)
    
    with open(output_path, 'w') as f:
        json.dump(related_docs_map, f, indent=2)


def print_stats(valid_hits, related_docs_map):
    """Print generation statistics"""
    docs_with_related = sum(1 for docs in related_docs_map.values() if docs)
    total_related = sum(len(docs) for docs in related_docs_map.values())
    avg_related = total_related / docs_with_related if docs_with_related > 0 else 0
    
    print(f"\n📊 Results:")
    print(f"  - Documents processed: {len(valid_hits)}")
    print(f"  - Documents with related content: {docs_with_related}")
    print(f"  - Average related docs per page: {avg_related:.1f}")
    print(f"  - Total relationships: {total_related}")


def print_examples(related_docs_map):
    """Print example relationships"""
    examples = [(k, v) for k, v in related_docs_map.items() if v][:3]
    if examples:
        print(f"\n🔍 Examples:")
        for doc_id, related in examples:
            print(f"\n  {doc_id}:")
            for r in related[:2]:
                print(f"    → {r['title']} ({r['similarity_score']}%, {r['doc_type']})")


def main():
    parser = argparse.ArgumentParser(description='Generate related docs from Algolia index')
    parser.add_argument('--algolia_app_id', required=True, help='Algolia application ID')
    parser.add_argument('--algolia_api_key', required=True, help='Algolia API key')
    parser.add_argument('--index_name', default='clickhouse', help='Algolia index name')
    parser.add_argument('--output', default='src/data/related-docs.json', help='Output JSON file path')
    
    args = parser.parse_args()
    
    print("🚀 Starting related docs generation...")
    
    # Initialize client and get all documents
    client = SearchClientSync(args.algolia_app_id, args.algolia_api_key)
    all_hits = get_all_docs(client, args.index_name)
    
    if not all_hits:
        print("❌ No documents found!")
        return
    
    # Filter to valid documents and show stats
    valid_hits = filter_valid_hits(all_hits)
    print(f"📖 Processing {len(valid_hits)} valid documents")
    show_doc_type_distribution(valid_hits)
    
    # Generate related docs mapping
    related_docs_map = generate_related_docs_map(valid_hits)
    
    # Save results
    save_results(related_docs_map, args.output)
    
    # Print final statistics and examples
    print_stats(valid_hits, related_docs_map)
    print_examples(related_docs_map)
    
    print(f"\n🎉 Done! Related docs saved to {args.output}")


if __name__ == '__main__':
    main()